libraries/chain/webassembly/eos-vm-oc/code_cache.cpp

```diff
@@ -0,0 +1,376 @@
+#include <fc/log/logger_config.hpp> //set_thread_name
+
+#include <eosio/chain/webassembly/eos-vm-oc/code_cache.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/config.hpp>
+#include <eosio/chain/webassembly/common.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/memory.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/eos-vm-oc.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/intrinsic.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/compile_monitor.hpp>
+#include <eosio/chain/exceptions.hpp>
+
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <sys/mman.h>
+#include <linux/memfd.h>
+
+#include "IR/Module.h"
+#include "IR/Validate.h"
+#include "WASM/WASM.h"
+#include "LLVMJIT.h"
+
+using namespace IR;
+using namespace Runtime;
+
+namespace eosio { namespace chain { namespace eosvmoc {
+
+static constexpr size_t header_offset = 512u;
+static constexpr size_t header_size = 512u;
+static constexpr size_t total_header_size = header_offset + header_size;
+static constexpr uint64_t header_id = 0x31434f4d56534f45ULL; //"EOSVMOC1" little endian
+
+struct code_cache_header {
+   uint64_t id = header_id;
+   bool dirty = false;
+   uintptr_t serialized_descriptor_index = 0;
+} __attribute__ ((packed));
+static constexpr size_t header_dirty_bit_offset_from_file_start = header_offset + offsetof(code_cache_header, dirty);
+static constexpr size_t descriptor_ptr_from_file_start = header_offset + offsetof(code_cache_header, serialized_descriptor_index);
+
+static_assert(sizeof(code_cache_header) <= header_size, "code_cache_header too big");
+
+code_cache_async::code_cache_async(const bfs::path data_dir, const eosvmoc::config& eosvmoc_config, const chainbase::database& db) :
+   code_cache_base(data_dir, eosvmoc_config, db),
+   _result_queue(eosvmoc_config.threads * 2),
+   _threads(eosvmoc_config.threads)
+{
+   wait_on_compile_monitor_message();
+
+   _monitor_reply_thread = std::thread([this]() {
+      fc::set_thread_name("oc-monitor");
+      _ctx.run();
+   });
+}
+
+code_cache_async::~code_cache_async() {
+   _compile_monitor_write_socket.shutdown(local::datagram_protocol::socket::shutdown_send);
```

Comment:
I don't think it's safe to access the socket from multiple threads without synchronization.